### DISCUSSION TITLE
Extracting Data into Columns

### DISCUSSION THREAD

**Discussion by chanpreet90** [jsvine/pdfplumber#778] [Ask for help with specific PDFs]
I have a pdf which has data in tabular format and has 6 columns but the columns are not separated by boundaries so when I extract the data using pdfplumber, all the data comes in one column only and I want in separate columns.

How could I do that?

> For your reference:
15/03/2021 RTGS-UTIBR52021031300662458-VIRENDER KUMAR 2,60,635.00 2,94,873.94Cr
11/03/2021 IMPS/P2A/107018040382/XXXXXXXXXX0980/trf 49,500.00 34,238.94Cr
11/03/2021 IMPS/P2A/107018771795/KINGDOMHOTELAND/trf 35,000.00 83,738.94Cr

**Comment by jsvine** [comment]
Hi @chanpreet90, unfortunately, it's difficult to help without access to the PDF. Could you attach it? I would also recommend trying the "text" strategies listed in the [table-extraction settings documentation](https://github.com/jsvine/pdfplumber#table-extraction-settings).

**Reply by chanpreet90** [reply]
here it is - https://we.tl/t-RuSjV2pXyh

**Reply by samkit-jain** [reply]
A word of **caution ⚠️** @chanpreet90 Whatever redaction tool you have used hasn't done a good job. While the sensitive information like the account number is not visible, it is still very much copyable. Do <kbd>Ctrl</kbd>+<kbd>A</kbd> on the PDF and you'll see all the invisible text. Since it is a bank statement, request you disable the WeTransfer link to prevent any misuse. Going forward, please use sophisticated tools like https://github.com/JoshData/pdf-redactor to redact/delete any sensitive content from a PDF file.

**Comment by samkit-jain** [comment]
@chanpreet90 If "text" strategy doesn't help and there are no vertical separators, you can use the "explicit_vertical_lines" strategy. In your case, the extraction strategy as 
```json
{
    "vertical_strategy": "lines",
    "horizontal_strategy": "lines",
    "explicit_vertical_lines": [80, 360, 450, 560, 680]
}
```
works fine. It gives the following output
![image](https://user-images.githubusercontent.com/15127115/208092590-cd13aa4f-931e-4dcb-a61d-dda328578266.png)

**Reply by chanpreet90** [reply]
Thank you so much for your prompt response. However, there is one more issue, which is that I've noticed that some values in the narration column appear in multiple lines, but those lines are not coming.

**Reply by samkit-jain** [reply]
Could you please share the code to reproduce it, the actual output and the expected output?

**Reply by chanpreet90** [reply]
I have used your table settings only and getting the following result :

![z](https://user-images.githubusercontent.com/86623814/208355726-47190910-0dd3-4703-bb6c-d3addc1f26c8.PNG)

In PDF document, I have some rows in which the content is in multiple lines, but only the first line is visible in the extracted file. I have highlighted the rows for your reference.
 
So is it possible to combine the content of multiple rows into a single row?

**Reply by samkit-jain** [reply]
@chanpreet90 It seems you are viewing this in an Excel file in the default row height. Increase the row height and you shall see the missing data in a new line in the same cell. If not, please check how you are exporting the data. Maybe there is a bug in your code because when I run `.extract_tables()`, I am getting the correct result
```
['DATE', 'NARRATION', 'CHQ.NO.', 'WITHDRAWAL(DR)', 'DEPOSIT(CR)', 'BALANCE(INR)']
['05/04/2021', 'IMPS/P2A/109407241841/XXXXXXXXXX2155/sector103', '', '32,820.00', '', '39,65,685.65Cr']
['03/04/2021', 'BY INST 420011 : MICR CLG (CTS)', '', '', '1,50,000.00', '39,98,505.65Cr']
['31/03/2021', 'BY INST 153187 : MICR CLG (CTS)', '', '', '4,00,000.00', '38,48,505.65Cr']
['29/03/2021', 'BY INST 14543 : MICR CLG (CTS)', '', '', '4,50,000.00', '34,48,505.65Cr']
['29/03/2021', 'BY INST 817608 : MICR CLG (CTS)', '', '', '9,25,000.00', '29,98,505.65Cr']
['29/03/2021', 'BY INST 751569 : MICR CLG (CTS)', '', '', '1,50,000.00', '20,73,505.65Cr']
['29/03/2021', 'BY INST 4692 : MICR CLG (CTS)', '', '', '2,00,000.00', '19,23,505.65Cr']
['29/03/2021', 'IMPS/P2A/108711200632/XXXXXXXXXX2155/sector103', '', '32,820.00', '', '17,23,505.65Cr']
['24/03/2021', 'INWARD RETURN', '', '177.00', '', '17,56,325.65Cr']
['24/03/2021', 'REJECT:187002:FUNDS INSUFFICIENT', '', '1,00,000.00', '', '17,56,502.65Cr']
['24/03/2021', 'BY INST 187002 : MICR CLG (CTS)', '', '', '1,00,000.00', '18,56,502.65Cr']
['24/03/2021', 'RTGS-BARB202103241083816876-KINGDOM HOTEL AND\nRESO', '', '5,72,530.00', '', '17,56,502.65Cr']
['24/03/2021', 'Charges for PORD Customer Payment :000604573904', '', '58.00', '', '23,29,032.65Cr']
['24/03/2021', 'BY INST 471 : MICR CLG (CTS)', '', '', '1,17,000.00', '23,29,090.65Cr']
['24/03/2021', 'BY INST 23578 : MICR CLG (CTS)', '', '', '3,00,000.00', '22,12,090.65Cr']
['24/03/2021', 'IMPS/P2A/108312080193/XXXXXXXXXX2155/sector103', '', '49,230.00', '', '19,12,090.65Cr']
['23/03/2021', 'INWARD RETURN', '', '295.00', '', '19,61,320.65Cr']
['23/03/2021', 'REJECT:4691:DRAWERS SIGNATURE DIFFERS', '', '2,00,000.00', '', '19,61,615.65Cr']
['23/03/2021', 'RTGS-ICICR12021032300384084-RAJESH\nSHARMA547,SARAS', '', '', '5,85,884.00', '21,61,615.65Cr']
['23/03/2021', 'BY INST 4691 : MICR CLG (CTS)', '', '', '2,00,000.00', '15,75,731.65Cr']
['23/03/2021', 'BY INST 120 : MICR CLG (CTS)', '', '', '4,93,861.00', '13,75,731.65Cr']
['23/03/2021', 'BY INST 358097 : MICR CLG (CTS)', '', '', '2,37,039.00', '8,81,870.65Cr']
['23/03/2021', 'BY INST 595944 : MICR CLG (CTS)', '', '', '2,86,800.00', '6,44,831.65Cr']
['22/03/2021', 'INWARD RETURN', '', '295.00', '', '3,58,031.65Cr']
['22/03/2021', 'REJECT:3795:OTHER REASONS (PLEASE SPECIFY)', '', '5,85,884.00', '', '3,58,326.65Cr']
['22/03/2021', 'BY INST 3795 : MICR CLG (CTS)', '', '', '5,85,884.00', '9,44,210.65Cr']
['22/03/2021', 'BY INST 27496 : MICR CLG (CTS)', '', '', '2,00,000.00', '3,58,326.65Cr']
['20/03/2021', 'EBANK:1251625312\\21030600154475\\GST', '', '700.00', '', '1,58,326.65Cr']
['20/03/2021', 'EBANK:1251601526/BHARTIAIRTELSE/462103003910043/\nC', '', '9,440.00', '', '1,59,026.65Cr']
['20/03/2021', 'EBANK:1251599342/http://www.airtel.in/11011454463', '', '8,846.22', '', '1,68,466.65Cr']
['20/03/2021', 'EBANK:1251593648/http://www.airtel.in/11011452336', '', '742.22', '', '1,77,312.87Cr']
['20/03/2021', 'EBANK:1251592544/BHARTIAIRTELSE/462103003909761/\nC', '', '770.03', '', '1,78,055.09Cr']
['20/03/2021', 'EBANK:1251591373/BHARTIAIRTELSE/462103003909733/\nC', '', '1,178.82', '', '1,78,825.12Cr']
['19/03/2021', 'IMPS/P2A/107814542235/XXXXXXXXXX2155/sector103', '', '32,820.00', '', '1,80,003.94Cr']
['16/03/2021', 'IMPS/P2A/107512091724/XXXXXXXXXX2155/trf', '', '32,820.00', '', '2,12,823.94Cr']
['15/03/2021', 'IMPS/P2A/107416641816/XXXXXXXXXX2155/DIESEL', '', '49,230.00', '', '2,45,643.94Cr']
['15/03/2021', 'RTGS-UTIBR52021031300662458-VIRENDER KUMAR\nKATHPAL', '', '', '2,60,635.00', '2,94,873.94Cr']
['11/03/2021', 'IMPS/P2A/107018040382/XXXXXXXXXX0980/trf', '', '49,500.00', '', '34,238.94Cr']
['11/03/2021', 'IMPS/P2A/107018771795/KINGDOMHOTELAND/trf', '', '', '35,000.00', '83,738.94Cr']
```
You can see the `\n` in the narration strings meaning that the table is getting extracted correctly.

**Reply by chanpreet90** [reply]
Oh, okay. I appreciate your assistance.

**Reply by chanpreet90** [reply]
One more question - What if there is no "\n" and want the output to be on the same line?

**Reply by samkit-jain** [reply]
If there is no `\n`, the output will be on the same line. `\n` represents a line break.

**Reply by chanpreet90** [reply]
Input - 
![z](https://user-images.githubusercontent.com/86623814/208603866-e49c6e83-db9c-4ee0-a667-42c9d25a747d.PNG)

Using extract_table() - 
![z3](https://user-images.githubusercontent.com/86623814/208604180-0e1706b9-5fe5-46f6-be91-af18305c45ea.PNG)

Converting into dataframe -
![z2](https://user-images.githubusercontent.com/86623814/208604336-aabdeab1-6cb0-42f6-8aef-8f851eaa54bf.PNG)

How to combine BALANCE and TRANSFER together like in the input image?

I hope I am able to make you understand my query.

**Reply by samkit-jain** [reply]
This is a different PDF from the one you shared earlier. Assuming you are using `lines` as the horizontal separator and getting this output, it most likely means that there are hidden lines in the statement. You can refer to https://github.com/jsvine/pdfplumber/issues/488#issuecomment-894077855 to know how to discard them.

**Reply by chanpreet90** [reply]
Yes, this PDF is different.
https://filetransfer.io/data-package/9pM0Upkg#link

**Reply by chanpreet90** [reply]
> @chanpreet90 If "text" strategy doesn't help and there are no vertical separators, you can use the "explicit_vertical_lines" strategy. In your case, the extraction strategy as
> 
> ```json
> {
>     "vertical_strategy": "lines",
>     "horizontal_strategy": "lines",
>     "explicit_vertical_lines": [80, 360, 450, 560, 680]
> }
> ```
> 
> works fine. It gives the following output ![image](https://user-images.githubusercontent.com/15127115/208092590-cd13aa4f-931e-4dcb-a61d-dda328578266.png)


Why is there one blank row appearing when I join pages?
